### PR TITLE
Added tests to verify the Cluster Status

### PR DESCRIPTION
**Pull request by marinakog**
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/13589/
this pull request does not include phase 3: domain.status for servers in cluster mirrored to cluster.status (not implemented yet)

### CODE REVIEW

**Review by xiancao — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 1]
```diff
@@ -275,6 +276,139 @@ void testAddReplaceClusterResource() {
...
+
+    // create and deploy cluster resource(s)
+    ClusterResource cluster = createClusterResource(
+        cluster1Res, cluster1Name, domainNamespace, replicaCount);
+    logger.info("Creating ClusterResource {0} in namespace {1}",cluster1Res, domainNamespace);
+    createClusterAndVerify(cluster);
+
+    ClusterResource cluster2 = createClusterResource(
+        cluster2Res, cluster2Name, domainNamespace, replicaCount);
+    logger.info("Creating ClusterResource {0} in namespace {1}",cluster2Res, domainNamespace);
```
[the remark this answers is not in the document]
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/domain/ClusterStatus.java [hunk 2]
```diff
@@ -145,6 +157,49 @@ public void setReplicasGoal(Integer replicasGoal) {
...
+
+  public void setConditions(List<ClusterCondition> conditions) {
+    this.conditions = conditions;
+  }
+
+  public Integer getObservedGeneration() {
+    return observedGeneration;
+  }
+
+  public ClusterStatus withObservedGeneration(int observedGeneration) {
```
[the remark this answers is not in the document]
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/domain/ClusterStatus.java]: done

File: integration-tests/src/test/java/oracle/weblogic/domain/ClusterCondition.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+import java.time.OffsetDateTime;
+
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+@ApiModel(
+    description = "ClusterCondition contains details for the current condition of this domain.")
```
[the remark this answers is not in the document]
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/domain/ClusterCondition.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 4]
```diff
@@ -275,6 +276,139 @@ void testAddReplaceClusterResource() {
...
+    createClusterAndVerify(cluster2);
+
+    createModelConfigMap(domainUid,configMapName);
+
```
[the remark this answers is not in the document]
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 5]
```diff
@@ -275,6 +276,139 @@ void testAddReplaceClusterResource() {
...
+    for (int i = 1; i <= replicaCount; i++) {
+      logger.info("Wait for managed pod {0} to be ready in namespace {1}",
+          managedServer1Prefix + i, domainNamespace);
+      checkPodReadyAndServiceExists(managedServer1Prefix + i, domainUid, domainNamespace);
+    }
+    logger.info("Check cluster resource status is mirror of domain.status");
+    assertDoesNotThrow(() -> {
+          testUntil(
+              clusterStatusMatchesDomain(domainUid, domainNamespace, cluster1Name), getLogger(),
+              "Cluster Resource status does not match domain.status");
```
[the remark this answers is not in the document]
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 6]
```diff
@@ -699,11 +835,44 @@ void testKubectlScaleClusterResource() {
...
+                  "Available", "True"), getLogger(),
+              "Cluster Resource status condition type does not match domain.status");
+          testUntil(clusterStatusConditionsMatchesDomain(domainUid, domainNamespace, cluster1Name,
+                  "Completed", "True"), getLogger(),
+              "Cluster Resource status condition type does not match domain.status");
+          testUntil(clusterStatusMatchesDomain(domainUid, domainNamespace, cluster2Name), getLogger(),
+              "Cluster Resource status does not match domain.status");
+          testUntil(clusterStatusConditionsMatchesDomain(domainUid, domainNamespace, cluster2Name,
+                  "Available", "True"), getLogger(),
+              "Cluster Resource status condition type does not match domain.status");
```
[the remark this answers is not in the document]
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 7]
```diff
@@ -699,11 +835,44 @@ void testKubectlScaleClusterResource() {
...
+          testUntil(clusterStatusConditionsMatchesDomain(domainUid, domainNamespace, cluster2Name,
+                  "Completed", "True"), getLogger(),
+              "Cluster Resource status condition type does not match domain.status");
```
[the remark this answers is not in the document]
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 8]
```diff
@@ -699,11 +835,44 @@ void testKubectlScaleClusterResource() {
...
     checkPodReadyAndServiceExists(managedPod2Prefix + "3", domainUid, domainNamespace);
     checkPodDoesNotExist(managedPod1Prefix + "3", domainUid, domainNamespace);
 
+    //verify status and conditions after scaling
+    assertDoesNotThrow(() -> {
+          testUntil(clusterStatusMatchesDomain(domainUid, domainNamespace, cluster2Name), getLogger(),
+              "Cluster Resource status does not match domain.status");
+          testUntil(clusterStatusConditionsMatchesDomain(domainUid, domainNamespace, cluster2Name,
+                  "Available", "True"), getLogger(),
+              "Cluster Resource status condition type does not match domain.status");
```
[the remark this answers is not in the document]
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 9]
```diff
@@ -699,11 +835,44 @@ void testKubectlScaleClusterResource() {
...
+          testUntil(clusterStatusConditionsMatchesDomain(domainUid, domainNamespace, cluster2Name,
+                  "Completed", "True"), getLogger(),
+              "Cluster Resource status condition type does not match domain.status");
```
[the remark this answers is not in the document]
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 10]
```diff
@@ -277,6 +278,137 @@ void testAddReplaceClusterResource() {
     deleteClusterCustomResourceAndVerify(cluster2Res,domainNamespace);
   }
 
+  /**
+   * Create WebLogic domain DR with domain level replica set to zero.
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:282]: domain DR  --> Domain Resource

**Review by xiancao — APPROVED** [review]

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java (hunk 10 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:282]: done

File: integration-tests/src/test/java/oracle/weblogic/domain/ClusterStatus.java [hunk 11]
```diff
@@ -186,6 +245,8 @@ public boolean equals(Object other) {
         .append(maximumReplicas, rhs.maximumReplicas)
         .append(minimumReplicas, rhs.minimumReplicas)
         .append(replicasGoal, rhs.replicasGoal)
+        .append(observedGeneration, rhs.observedGeneration)
+        .append(conditions, rhs.conditions)
```
Comment (lennyphan) [inline integration-tests/src/test/java/oracle/weblogic/domain/ClusterStatus.java:249]: We're sorting the conditions for ClusterStatus (also check DomainStatus):

.append(DomainResource.sortList(conditions), DomainResource.sortList(rhs.conditions))

Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/domain/ClusterStatus.java]: done

**Pull request merged**